File: wire-android-automation/common/src/main/java/com/wearezeta/auto/common/email/inbucket/InbucketClient.java
/*
Copyright 2016 StepStone Services

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

package com.wearezeta.auto.common.email.inbucket;

import com.wearezeta.auto.common.email.inbucket.models.Message;
import com.wearezeta.auto.common.email.inbucket.models.MessageInfo;
import com.wearezeta.auto.common.log.ZetaLogger;
import okhttp3.*;
import retrofit2.Retrofit;
import retrofit2.converter.jackson.JacksonConverterFactory;

import java.io.IOException;
import java.net.Authenticator;
import java.net.InetSocketAddress;
import java.net.PasswordAuthentication;
import java.net.Proxy;
import java.util.List;
import java.util.logging.Logger;

public class InbucketClient {

    private static final Logger log = ZetaLogger.getLog(InbucketClient.class.getSimpleName());

    protected InbucketService service;

    protected InbucketService simpleService;

    public InbucketClient(String baseUrl, String user, String password, boolean useProxy) {
        OkHttpClient.Builder clientBuilder = new OkHttpClient().newBuilder();

        if (useProxy) {
            Authenticator authenticator = new Authenticator() {
                public PasswordAuthentication getPasswordAuthentication() {
                    return (new PasswordAuthentication("qa",
                            com.wearezeta.auto.common.credentials.Credentials.get("SOCKS_PROXY_PASSWORD").toCharArray()));
                }
            };
            Authenticator.setDefault(authenticator);
            Proxy proxy = new Proxy(Proxy.Type.SOCKS, new InetSocketAddress("socks.wire.link", 1080));
            clientBuilder.proxy(proxy);
        }

        class LoggingInterceptor implements Interceptor {
            @Override public Response intercept(Chain chain) throws IOException {
                Request request = chain.request();

                long t1 = System.nanoTime();
                log.info(String.format("Sending request %s on %s%n%s", request.url(), chain.connection(),
                        request.headers()));

                Response response = chain.proceed(request);

                long t2 = System.nanoTime();
                log.info(String.format("Received response for %s in %.1fms%n%s", response.request().url(),
                        (t2 - t1) / 1e6d, response.headers()));

                return response;
            }
        }

        OkHttpClient okHttpClient = clientBuilder
                .addInterceptor(chain -> {
                    Request originalRequest = chain.request();

                    Request.Builder builder = originalRequest.newBuilder().header("Authorization",
                            Credentials.basic(user, password));

                    Request newRequest = builder.build();
                    return chain.proceed(newRequest);
                })
                .addInterceptor(new LoggingInterceptor())
                .build();

        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(okHttpClient)
                .addConverterFactory(JacksonConverterFactory.create())
                .build();
        service = retrofit.create(InbucketService.class);
        retrofit = new Retrofit.Builder()
                .baseUrl(baseUrl)
                .client(okHttpClient)
                .build();
        simpleService = retrofit.create(InbucketService.class);
    }

    /**
     *  Method callss inbuckets GET /api/v1/mailbox/{mailboxName} endpoint
     * @param mailboxName name of mailbox
     * @return List of MailboxEntry
     * @throws IOException if something went wrong
     */
    public List<MessageInfo> getMailbox(String mailboxName) throws IOException {
        return service.getMailbox(mailboxName).execute().body();
    }

    /**
     *  Method calls inbuckets GET /api/v1/mailbox/{mailboxName}/{messageId} endpoint
     * @param mailboxName name of mailbox
     * @param messageId id of message
     * @return List of Message
     * @throws IOException if something went wrong
     */
    public Message getMessage(String mailboxName, String messageId) throws IOException {
        return service.getMessage(mailboxName, messageId).execute().body();
    }

    /**
     * Method calls inbuckets GET /api/v1/mailbox/{mailboxName}/{messageId}/source endpoint
     * @param mailboxName name of mailbox
     * @param messageId id of message
     * @return message source as string
     * @throws IOException if something went wrong
     */
    public String getMessageSource(String mailboxName, String messageId) throws IOException {
        return simpleService.getMessageSource(mailboxName, messageId).execute().body().string();
    }


    /**
     * Method calls inbuckets DELETE /api/v1/mailbox/{mailboxName} endpoint
     * @param mailboxName name of mailbox
     * @return "OK" string on success
     * @throws IOException on error
     */
    public String deleteMailbox(String mailboxName) throws IOException {
             return service.deleteMailbox(mailboxName).execute().body();
    }

    /**
     *  Method calls inbuckets DELETE /api/v1/mailbox/{mailboxName}/{messageId} endpoint
     * @param mailboxName name of mailbox
     * @param messageId id of message
     * @return "OK" string on success null on error
     * @throws IOException on error
     */
    public String deleteMessage(String mailboxName, String messageId) throws IOException {
        return service.deleteMessage(mailboxName,messageId).execute().body();
    }
}

File: tests/testsCore/src/main/kotlin/com/wire/android/tests/core/models/InbucketClient.kt

package com.wire.android.tests.core.models

import com.wire.android.tests.core.utils.ZetaLogger
import okhttp3.Credentials
import okhttp3.Interceptor
import okhttp3.OkHttpClient
import okhttp3.Response
import retrofit2.Retrofit
import retrofit2.converter.jackson.JacksonConverterFactory
import java.io.IOException
import java.net.Authenticator
import java.net.InetSocketAddress
import java.net.PasswordAuthentication
import java.net.Proxy
import java.util.logging.Logger

class InbucketClient(baseUrl: String?, user: String, password: String, useProxy: Boolean) {

    companion object {
        private val log: Logger = ZetaLogger.getLog(InbucketClient::class.simpleName)
    }

    protected val service: InbucketService
    protected val simpleService: InbucketService

    init {
        val clientBuilder = OkHttpClient().newBuilder()

        if (useProxy) {
            Authenticator.setDefault(object : Authenticator() {
                override fun getPasswordAuthentication(): PasswordAuthentication {
                    val credentials = com.wire.android.tests.core.config.Credentials()
                    return PasswordAuthentication(
                        "qa",
                        credentials.getCredentials("SOCKS_PROXY_PASSWORD").toCharArray()
                    )
                }
            })
            val proxy = Proxy(Proxy.Type.SOCKS, InetSocketAddress("socks.wire.link", 1080))
            clientBuilder.proxy(proxy)
        }

        class LoggingInterceptor : Interceptor {
            @Throws(IOException::class)
            override fun intercept(chain: Interceptor.Chain): Response {
                val request = chain.request()

                val t1 = System.nanoTime()
                log.info("Sending request ${request.url} on ${chain.connection()}\n${request.headers}")

                val response = chain.proceed(request)

                val t2 = System.nanoTime()
                log.info("Received response for ${response.request.url} in ${(t2 - t1) / 1e6}ms\n${response.headers}")

                return response
            }
        }

        val okHttpClient = clientBuilder
            .addInterceptor { chain ->
                val originalRequest = chain.request()
                val newRequest = originalRequest.newBuilder()
                    .header("Authorization", Credentials.basic(user, password))
                    .build()
                chain.proceed(newRequest)
            }
            .addInterceptor(LoggingInterceptor())
            .build()

        service = Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(okHttpClient)
            .addConverterFactory(JacksonConverterFactory.create())
            .build()
            .create(InbucketService::class.java)

        simpleService = Retrofit.Builder()
            .baseUrl(baseUrl)
            .client(okHttpClient)
            .build()
            .create(InbucketService::class.java)
    }

    @Throws(IOException::class)
    fun getMailbox(mailboxName: String): List<MessageInfo> =
        service.getMailbox(mailboxName).execute().body()!!

    @Throws(IOException::class)
    fun getMessage(mailboxName: String, messageId: String): Message =
        service.getMessage(mailboxName, messageId).execute().body()!!

    @Throws(IOException::class)
    fun getMessageSource(mailboxName: String, messageId: String): String =
        simpleService.getMessageSource(mailboxName, messageId).execute().body()!!.string()

    @Throws(IOException::class)
    fun deleteMailbox(mailboxName: String): String =
        service.deleteMailbox(mailboxName).execute().body()!!

    @Throws(IOException::class)
    fun deleteMessage(mailboxName: String, messageId: String): String? =
        service.deleteMessage(mailboxName, messageId).execute().body()
}
